### PR TITLE
Expose String.EndsWith(char) and String.StartsWith(char)

### PR DESCRIPTION
**Pull request by fujiy**
Expose String.EndsWith(char) and String.StartsWith(char). Depends on https://github.com/dotnet/coreclr/pull/9228

### CODE REVIEW

**Comment by justinvp** [comment]
We'll need tests for the new APIs as well.

**Comment by danmosemsft** [comment]
That's correct, thanks @justinvp for pointing that out.

**Comment by fujiy** [comment]
@juntinvp how do you test (the tests are in CoreFx) your own coreclr build? The docs https://github.com/dotnet/coreclr/blob/master/Documentation/project-docs/adding_new_public_apis.md says to wait the CoreClr changes be built.

Before the CoreFx changes from project.json I could change the dependency version and add my local coreclr folder as NuGet repository.

Thanks

**Comment by justinvp** [comment]
It's been a while since I've tested new APIs with a locally-built CoreCLR, so I'm not sure how up-to-date this is, but see https://github.com/dotnet/coreclr/blob/master/Documentation/building/testing-with-corefx.md

**Comment by danmosemsft** [comment]
I think that's out of date, we should update it. This should work
https://github.com/dotnet/corefx/blob/master/Documentation/project-docs/developer-guide.md#testing-with-private-coreclr-bits

File: src/System.Runtime/tests/System/StringTests.cs [hunk 1]
```diff
@@ -816,6 +816,22 @@ public static void EndsWith(string s, string value, StringComparison comparisonT
         }
 
         [Theory]
+        [InlineData("Hello", 'o', true)]
+        [InlineData("Hello", 'O', false)]
+        [InlineData("o", 'o', true)]
+        [InlineData("o", 'O', true)]
```
Comment (danmosemsft) [inline src/System.Runtime/tests/System/StringTests.cs]: Why is this true? I believe we want to match .EndsWith(string) behavior which is case senseitive

https://msdn.microsoft.com/en-us/library/2333wewz(v=vs.110).aspx

Comment (danmosemsft) [inline src/System.Runtime/tests/System/StringTests.cs]: Oops, it looks like they ended up wanting to match .IndexOf(char) behavior. That is still case sensitive -it's ordinal.

https://msdn.microsoft.com/en-us/library/kwb0bwyd(v=vs.110).aspx

**Comment by danmosemsft** [comment]
LGTM with the tweak I noted to fix the test.

**Comment by danmosemsft** [comment]
If you wanted, you could generate the "expected" value yourself by having the test do `s.LastIndexOf('c') == s.Length - 1`

**Comment by danmosemsft** [comment]
https://github.com/dotnet/corefx/issues/4805

**Comment by danmosemsft** [comment]
The CI will fail until it ingests the CLR change of course.

**Comment by fujiy** [comment]
test all

**Comment by fujiy** [comment]
@dotnet-bot test Innerloop Windows_NT Release Build and Test

**Comment by fujiy** [comment]
@dotnet-bot test

**Comment by fujiy** [comment]
@dotnet-bot test Innerloop Windows_NT Debug Build and Test

**Comment by fujiy** [comment]
@dotnet-bot test Innerloop CentOS7.1 Debug Build and Test
@dotnet-bot test Innerloop CentOS7.1 Release Build and Test

**Comment by fujiy** [comment]
@dotnet-bot test Innerloop OSX Debug Build and Test — Build finished
@dotnet-bot test Innerloop OSX Release Build and Test — Build finished
@dotnet-bot test Innerloop PortableLinux Debug Build and Test — Build finished
@dotnet-bot test Innerloop PortableLinux Release Build and Test — Build finished
@dotnet-bot test Innerloop Ubuntu14.04 Debug Build and Test — Build finished
@dotnet-bot test Innerloop Ubuntu14.04 Release Build and Test — Build finished
@dotnet-bot test Vertical uap Build — Build finished
@dotnet-bot test Vertical uapaot Build — Build finished
@dotnet-bot test Windows_NT Debug AllConfigurations Build — Build finished

**Comment by fujiy** [comment]
Does UAP use CoreRT? We probably need to implement those methods in CoreRT repo.

**Comment by danmosemsft** [comment]
@joperezr any idea why these two builds specifically would fail with 'missing implementation' but not the others? The implementation is in corelib, it should not be configuration specific.

**Comment by danmosemsft** [comment]
Oh -- it's as @fujiy says -- allconfigurations builds aot as well, and it's missing from the corert corelib. Is this going to happen from now on (now we have aot in CI) every time we consume new corelib API?

**Comment by joperezr** [comment]
> @joperezr any idea why these two builds specifically would fail with 'missing implementation' but not the others? The implementation is in corelib, it should not be configuration specific.

The build that's failing is the uapaot one and it is because NETNative's corelib doesn't have these new exposed API. To fix this build, simply run: `msbuild src\System.Runtime\src\System.Runtime.csproj /p:TargetGroup=uapaot /p:BaselineAllApiCompatError=true`

**Comment by danmosemsft** [comment]
Ungh -- of course -- I forgot the baseline, @fujiy please update the baseline as shown and update the PR.

If you're willing to update CoreRT afterwards, great!

**Comment by fujiy** [comment]
@danmosemsft, @joperezr I ran the command, it updated the src/System.Runtime/src/ApiCompatBaseline.uapaot.txt file, including things not related to my changes.
Do we have any docs about how it works? I was not aware of this and would like to learn more.
Thanks!

**Comment by joperezr** [comment]
> Do we have any docs about how it works? I was not aware of this and would like to learn more.
Thanks!

We don't really have docs about this since it is a point in time thing. Eventually, NetNative's CoreLib will have the types that were moved down to CoreCLR's CoreLib, so type-forwarding on uapaot will still work. What was happening is that with your change you were saying that those types existed in CoreLib and should by typeforwarded down, but NETNative doesn't  have them, so our tool called ApiCompat was flagging that. By passing that extra argument, we tell ApiCompat to just baseline the errors for now since they are expected, and will only be fixed until we actually have the types down there. If you are interested on how our ApiCompat tool works, it is open sourced here: https://github.com/dotnet/buildtools/tree/master/src/ApiCompat

**Comment by danmosemsft** [comment]
Merged. Thank you for your work on this @fujiy . I hope it hasn't put you off making other contributions - they're usually not this time consuming to get through.

**Comment by fujiy** [comment]
@danmosemsft not at all, it just made it more interesting. I have talked with @karelz before, about the plan I made with the my colleagues. We are contributing to some .NET issues to learn more about these new technologies and it has been worth. If everything goes well, we can continue and expand the plan.
I'll keep you in touch

**Pull request merged**